src/main/java/net/dries007/tfc/common/container/KnappingContainer.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Licensed under the EUPL, Version 1.2.
+ * You may obtain a copy of the Licence at:
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ */
+
+package net.dries007.tfc.common.container;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.PlayerInventory;
+import net.minecraft.inventory.container.ContainerType;
+import net.minecraft.inventory.container.Slot;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.IRecipeType;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.util.Hand;
+import net.minecraft.util.SoundEvent;
+import net.minecraft.world.server.ServerWorld;
+
+import net.minecraftforge.items.ItemHandlerHelper;
+import net.minecraftforge.items.ItemStackHandler;
+
+import net.dries007.tfc.common.TFCTags;
+import net.dries007.tfc.common.recipes.IInventoryNoop;
+import net.dries007.tfc.common.recipes.SimpleCraftMatrix;
+import net.dries007.tfc.common.recipes.KnappingRecipe;
+import net.dries007.tfc.util.Helpers;
+
+public class KnappingContainer extends ItemStackContainer implements IButtonHandler, IInventoryNoop
+{
+    public final SimpleCraftMatrix matrix;
+    private final int amountToConsume;
+    public final boolean usesDisabledTex;
+    private final boolean needsKnife;
```

Comment:
This is really specific and not very useful to extensions. Rather than this, add a subclass of KnappingContainer used for leather knapping which overrides just `consumeIngredientStackAfterComplete`. That's better than declaring knives as an inherent part of knapping itself.